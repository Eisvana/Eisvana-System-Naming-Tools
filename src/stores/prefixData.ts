import { useCoords2Glyphs } from '@/composables/coordGlyphConvert';
import { defineStore } from 'pinia';

interface State {
  glyphs: string;
  systemName: string;
  prefix: string;
}

export const usePrefixDataStore = defineStore('prefixData', {
  state: (): State => ({
    glyphs: '',
    systemName: '',
    prefix: '',
  }),

  getters: {
    glyphValues: (state) => useCoords2Glyphs(state.glyphs),
  },
});
