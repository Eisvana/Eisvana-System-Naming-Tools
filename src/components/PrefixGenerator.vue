<script setup lang="ts">
import GlyphInput from './GlyphInput.vue';
import { maxGlyphLength } from '../variables/constants';
import { computed } from 'vue';

import { usePrefixDataStore } from '../stores/prefixData';
import { storeToRefs } from 'pinia';
import { getPrefix, getRegionNumber } from '@/logic/prefix';

const prefixDataStore = usePrefixDataStore();
const { glyphs, glyphValues, systemName } = storeToRefs(prefixDataStore);

const prefix = computed(() => {
  const barePrefix = getPrefix(glyphValues.value);
  return barePrefix ? `${barePrefix} ${systemName.value}` : '';
});

function resetTagGenerator() {
  glyphs.value = '';
  systemName.value = '';
}

const outputLabel = computed(() => (systemName.value ? 'System Name' : 'Prefix'));

const isNotEisvana = computed(() => !getRegionNumber(glyphValues.value) && glyphValues.value.length === maxGlyphLength);
</script>

<template>
  <h2>Name Generator</h2>
  <p>A proper system name consists of two parts:</p>
  <div class="input-group">
    <div>
      <label for="systemName">System Name:</label>
      <input
        id="systemName"
        type="text"
        v-model="systemName"
      />
    </div>
    <GlyphInput
      class="glyph-input"
      :error="isNotEisvana"
    />
  </div>
  <div
    class="status-wrapper"
    v-show="isNotEisvana"
  >
    <p class="status error">Not in Eisvana!</p>
  </div>
  <div
    v-show="prefix"
    class="output-wrapper"
  >
    <p class="output-label">{{ outputLabel }}:</p>
    <output class="prefix-output">{{ prefix.trim() }}</output>
  </div>
  <button
    v-show="glyphs || systemName"
    class="reset-button secondary"
    type="reset"
    @click="resetTagGenerator"
  >
    Reset
  </button>
</template>

<style scoped>
.glyph-input {
  flex-grow: 1;
  margin-block-end: 1rem;
}
</style>
