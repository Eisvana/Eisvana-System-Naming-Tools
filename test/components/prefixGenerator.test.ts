import PrefixGenerator from '@/components/PrefixGenerator.vue';
import { beforeEach, describe, expect, it } from 'vitest';
import { mount, type VueWrapper } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { usePrefixDataStore } from '@/stores/prefixData';
import { storeToRefs } from 'pinia';

describe('PrefixGenerator', () => {
  let wrapper: VueWrapper;
  const pinia = createTestingPinia();

  beforeEach(() => {
    wrapper = mount(PrefixGenerator, {
      global: {
        plugins: [pinia],
      },
    });
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should reset data on button click', async () => {
    const store = usePrefixDataStore(pinia);
    const { glyphs, systemName } = storeToRefs(store);
    glyphs.value = '0001A21117FF';
    systemName.value = 'Hello World';
    const button = wrapper.find('button.reset-button');
    await button.trigger('click');

    expect(glyphs.value + systemName.value).toBe('');
  });

  it('should have the correct label', async () => {
    const output = wrapper.find('.output-label');

    expect(output.element.textContent).toBe('Prefix:');
  });

  it('should have the correct label', async () => {
    const systemInput = wrapper.find('#systemName');
    const glyphInput = wrapper.find('#portalglyphsInput');
    const output = wrapper.find('.prefix-output');
    await systemInput.setValue('Hello World');
    await glyphInput.setValue('0001A21117FF');

    expect(output.element.textContent).toBe('EV1-1 Hello World');
  });
});
