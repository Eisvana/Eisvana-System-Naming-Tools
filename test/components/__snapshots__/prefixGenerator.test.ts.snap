// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrefixGenerator > should render correctly 1`] = `
"<h2 data-v-49337c37="">Name Generator</h2>
<p data-v-49337c37="">A proper system name consists of two parts:</p>
<div data-v-49337c37="" class="input-group">
  <div data-v-49337c37=""><label data-v-49337c37="" for="systemName">System Name:</label><input data-v-49337c37="" id="systemName" type="text"></div>
  <div data-v-ed03a538="" data-v-49337c37="" class="glyph-input"><label data-v-ed03a538="" class="required" for="portalglyphsInput">Glyphs or Coordinates:</label>
    <div data-v-ed03a538="" class="glyph-input-wrapper"><input data-v-ed03a538="" maxlength="19" class="glyphs-input" id="portalglyphsInput" type="text"><button data-v-ed03a538="" class="delete-button is-error" id="delButton" role="button" type="button"> ‚Üê Delete </button></div>
    <div data-v-ed03a538="" class="portal-buttons grid"><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton1" value="0">0</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton2" value="1">1</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton3" value="2">2</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton4" value="3">3</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton5" value="4">4</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton6" value="5">5</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton7" value="6">6</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton8" value="7">7</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton9" value="8">8</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton10" value="9">9</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton11" value="A">A</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton12" value="B">B</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton13" value="C">C</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton14" value="D">D</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton15" value="E">E</button><button data-v-ed03a538="" class="button glyphs" type="button" id="glyphButton16" value="F">F</button></div>
    <p data-v-ed03a538="" class="glyph-display-wrapper" style="display: none;"><output data-v-ed03a538="" class="glyphs" id="glyphDisplay"></output></p>
  </div>
</div>
<div data-v-49337c37="" class="status-wrapper" style="display: none;">
  <p data-v-49337c37="" class="status error">Not in Eisvana!</p>
</div>
<div data-v-49337c37="" class="output-wrapper" style="display: none;">
  <p data-v-49337c37="" class="output-label">Prefix:</p><output data-v-49337c37="" class="prefix-output"></output>
</div>
<button data-v-49337c37="" class="reset-button secondary" type="reset" style="display: none;"> Reset </button>"
`;
